compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/ProvidesMethodFactoryGenerator.kt

```diff
@@ -293,4 +289,28 @@ internal class ProvidesMethodFactoryGenerator : PrivateCodeGenerator() {
 
     return createGeneratedFile(codeGenDir, packageName, className, content)
   }
+
+  private fun checkFunctionIsNotAbstract(
+    clazz: KtClassOrObject,
+    function: KtFunction
+  ) {
+    fun fail(): Nothing = throw AnvilCompilationException(
+      "@Provides methods cannot be abstract",
+      element = function
+    )
+
+    // If the function is abstract, then it's an error.
+    if (function.hasModifier(ABSTRACT_KEYWORD)) fail()
```

Comment:
This method has a bit of code smell because of how the fail states operate as sort of a state machine. Would it be better to split out each of these checks into it's own method instead? Because checkFunctionIsNotAbstract doesn't really capture all that's going on here.

Reply:
Splitting might not be the best either. The flow just seems odd. in it's current form.

Reply:
It's not a state machine but rather a control flow. I don't think splitting would bring any value. I also believe that the method name describes well what's going on. All functions in interfaces are abstract by default and all functions with the `abstract` keyword are abstract. This method checks for both of these two cases ¯\_(ツ)_/¯